### PR TITLE
bauhaus: add callback feature to bauhaus slider

### PR DESCRIPTION
**Pull request by upegelow**
A callback function can be defined to enable a transparent conversion
between the linear slider position and a potentially non-linear outer
value. Typical use case is a logarithmic slider scale which can be defined
by a callback function as:

```
static float log_callback(float inval, dt_bauhaus_callback_t dir)
{
  float outval;
  switch(dir)
  {
    case DT_BAUHAUS_SET:
      outval = log10(fmax(inval, 1e-15f));
      break;
    case DT_BAUHAUS_GET:
      outval = exp(M_LN10 * inval);
      break;
    default:
      outval = inval;
  }
  return outval;
}
```

Note that after a callback is installed all interface functions that set or get
absolute slider values will go through the callback. Exception:
dt_bauhaus_slider_get_step() which returns a delta in linear coordinates.

### CODE REVIEW

**Comment by upegelow** [comment]
If there are no objections I'll merge soon.

File: src/bauhaus/bauhaus.c [hunk 1]
```diff
@@ -714,7 +715,7 @@ float dt_bauhaus_slider_get_hard_min(GtkWidget* widget)
 {
   dt_bauhaus_widget_t *w = DT_BAUHAUS_WIDGET(widget);
   dt_bauhaus_slider_data_t *d = &w->data.slider;
-  return d->hard_min;
+  return (d->callback == NULL ? d->hard_min : d->callback(d->hard_min, DT_BAUHAUS_GET));
```
Comment (LebedevRI) [inline src/bauhaus/bauhaus.c]: I would just add a linear default callback.
Would save one the trouble of thinking whether there is a callback.

**Comment by boucman** [comment]
hmm, would it be possible to add a "self" parameter to this ?

I might want to add a way for lua to implement that function and lua code need to have access to the widget itself to know what it's working on...

File: src/bauhaus/bauhaus.c (hunk 1 above)
Comment (upegelow) [inline src/bauhaus/bauhaus.c]: OTOH it's more simple now to reset to default by just dt_bauhaus_slider_set_callback(widget, NULL).

**Comment by upegelow** [comment]
> hmm, would it be possible to add a "self" parameter to this ?

How should this look like?

**Comment by boucman** [comment]
adding a "GtkWidget* self" parameter to the callback pointing to the widget for whom we are calculating the value would be enough for my need...

**Comment by upegelow** [comment]
Please check if this is what you want. Callback needs to be defined now as

float callback(GktWidget *self, float inval, dt_bauhaus_callback_t dir)

**Comment by boucman** [comment]
yes, that's it, thx a lot

**Comment by upegelow** [comment]
You're welcome :)

**Pull request merged**